python/adelphi/adelphi/store.py

```diff
@@ -0,0 +1,87 @@
+# Functions to facilitate interactions with the underlying data store
+
+import logging
+from itertools import tee
+
+# Account for name change in itertools as of py3k
+try:
+    from itertools import ifilterfalse as filterfalse
+except ImportError:
+    from itertools import filterfalse
+
+from cassandra.cluster import Cluster, ExecutionProfile, EXEC_PROFILE_DEFAULT, default_lbp_factory
+from cassandra.auth import PlainTextAuthProvider
+
+logging.basicConfig(level=logging.INFO)
+log = logging.getLogger('adelphi')
+
+system_keyspaces = set(["system",
+                    "system_schema",
+                    "system_traces",
+                    "system_auth",
+                    "system_distributed",
+                    "system_virtual_schema",
+                    "system_views"])
+
+def build_auth_provider(username = None,password = None):
+    # instantiate auth provider if credentials have been provided
+    auth_provider = None
+    if username is not None and password is not None:
+        auth_provider = PlainTextAuthProvider(username=username, password=password)
+    return auth_provider
+
+
+def with_cluster(cluster_fn, hosts, port, username = None, password = None):
+    ep = ExecutionProfile(load_balancing_policy=default_lbp_factory())
+    cluster = Cluster(hosts, port=port, auth_provider=build_auth_provider(username,password), execution_profiles={EXEC_PROFILE_DEFAULT: ep})
+    log.info("Connecting to the cluster to get metadata...")
+    cluster.connect()
+    cluster_fn(cluster)
+    cluster.shutdown()
+
+
+def build_keyspace_objects(keyspaces, metadata):
+    """Build a list of cassandra.metadata.KeyspaceMetadata objects from a list of strings and a c.m.Metadata instance.  System keyspaces will be excluded."""
+    all_keyspace_objs = [metadata.keyspaces[ks] for ks in keyspaces] if keyspaces is not None else metadata.keyspaces.values()
+
+    # Borrowed from itertools
+    def partition(pred, iterable):
+        t1, t2 = tee(iterable)
+        return list(filterfalse(pred, t1)), list(filter(pred, t2))
+
+    (failed,passed) = partition(lambda ks: ks.name not in system_keyspaces,all_keyspace_objs)
+    if failed:
+        log.info("Excluding system keyspaces " + ",".join((ks.name for ks in failed)))
```

Comment:
It seemed worthwhile to be _very_ explicit here.  It's especially true if we're modifying a list of keyspaces explicitly supplied by the user but even in the "include all keyspaces" case it's worth being explicit here.